lib/oli/delivery/attempts/activity_lifecycle/persistence.ex

```diff
@@ -86,6 +85,24 @@ defmodule Oli.Delivery.Attempts.ActivityLifecycle.Persistence do
     end
   end
 
+  def bulk_update_activity_attempts(_, []), do: {:ok, []}
+
+  def bulk_update_activity_attempts(values, params) do
+    sql = """
+      UPDATE activity_attempts
+      SET
+        score = batch_values.score,
+        out_of = batch_values.out_of,
+        lifecycle_state = batch_values.lifecycle_state,
+        date_evaluated = batch_values.date_evaluated,
+        date_submitted = batch_values.date_submitted
+      FROM (VALUES #{values}) AS batch_values (activity_attempt_guid, score, out_of, lifecycle_state, date_evaluated, date_submitted)
+      WHERE activity_attempts.attempt_guid = batch_values.activity_attempt_guid
+    """
+
+    Ecto.Adapters.SQL.query(Oli.Repo, sql, params)
```

Comment:
I ask you to have a better understanding, this could have been written using Ecto expressions, right? Is there a more efficient way?

Reply:
I'm unaware of a way to do bulk record updates directly with Ecto.  This `batch_values` approach here is used in a few other places within Torus.  It might be worthwhile to eventually create a "builder" module for more easily creating and maintaining these types of queries.  Something that takes this out of string literal space back into Elixir space. I could imagine something like:

```
BulkUpdate.update(:activity_attempts)
|> BulkUpdate.set([:score, :out_of, :lifecycle_state, :date_evaluated, :date_submitted])
|> BulkUpdate.from_values(values)
|> BulkUpdate.matching_on(:attempt_guid)
```

Not for this PR, but perhaps in the future.

Reply:
Like it! Created this ticket to track the change for the future: https://eliterate.atlassian.net/browse/MER-1665